fix: avoid overhead when calling MediaQuery._init() multiple times

SassNinja opened the pull request:
<!------------------------------------------------------------------------------
                    Please fill the following template.
            For more information, see the CONTRIBUTING.md document
------------------------------------------------------------------------------->

## Description
<!-------------------------------------------------------------------
│   Describe your changes in detail. Include any relevant information:
│   reasons, difficulties, links to web references, related issues...
└------------------------------------------------------------------->

This makes sure Foundation.MediaQuery.queries doesn't contain several duplicates as it gets called by many plugins.


<!-------------------------------------------------------------------
│   Bugs and new features/improvements must be presented and
│   discussed in an issue first. Please create one if there is no issue
│   related to this pull request. You can skip this step for minor changes.
└------------------------------------------------------------------->
- Closes https://github.com/zurb/foundation-sites/issues/11540


## Types of changes
<!-------------------------------------------------------------------
│   What types of changes does your code introduce?
│   Fill with [x] all the boxes that apply:
└------------------------------------------------------------------->
- [ ] Documentation
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (anything that would change an existing functionality)
- [ ] Maintenance (refactor, code cleaning, development tools...)


## Checklist
<!-------------------------------------------------------------------
│   Please ensure that all the following points are respected.
│   Fill with [x] the boxes once the rule is respected.
└------------------------------------------------------------------->
- [x] I have read and follow the CONTRIBUTING.md document.
- [x] The pull request title and template are correctly filled.
- [x] The pull request targets the right branch (`develop` or `develop-v...`).
- [x] My commits are correctly titled and contain all relevant information.
- [ ] I have updated the documentation accordingly to my changes (if relevant).
- [ ] I have added tests to cover my changes (if relevant).


<!------------------------------------------------------------------------------
            For more information, see the CONTRIBUTING.md document         
              Thank you for your pull request and happy coding ;)          
------------------------------------------------------------------------------->

SassNinja committed:
fix: avoid overhead when calling MediaQuery._init() multiple times

This makes sure Foundation.MediaQuery.queries doesn't contain several duplicates because it gets called by many plugins.

ncoden reviewed (commented):
Thank you @SassNinja for the pull request 👍,

What if the CSS configuration is not availaible at the first initialization and appear later? It would not work anymore with a single initialization.

We could fix #11540 by reloading `queries` instead.

What do you think ?

ncoden commented:
> Usually the CSS should be ready before the JS but I guess it may be a problem if loading CSS via dynamic imports.

_Usually_ 😉 

> So you would prefer to reset the queries before populating it (again)?

Yes, this or adding a method to manually reinitialize MediaQuery. Do you have any objection ?

SassNinja commented:
> Yes, this or adding a method to manually reinitialize MediaQuery. Do you have any objection?

A) reset

- pro: impossible to break any existing project
- con: not performance optimized (meta elem gets added and `parseStyleToObject`  executed multiple times)

B) _reinit method

- pro: better performance (no unnecessary reinitialization)
- con: might break existing projects if the developer wasn't aware his CSS is loaded too late (he has to fix it himself with the new _reinit method)

I think I'd go with option B because it's more important to be as fast as possible instead of slowing everything down due to rare cases (CSS loaded too late).
We can include this in the release notes so devs know what's wrong when they are facing MediaQuery issues after updating.

Do you agree?

SassNinja committed:
fix: add _reinit method to cover case when CSS is loaded too late

ncoden commented:
Possible breaking change. Planned for v6.6.0

ncoden commented on js/foundation.util.mediaQuery.js:
```diff
@@ -105,6 +113,16 @@ var MediaQuery = {
+   */
+  _reinit() {
+    this.isInitialized = false;
+    this._init();
```
This would populate the `queries` array again.
You could remove `queries: []` and set up `queries` in the `_init()` method.

ncoden commented on js/foundation.util.mediaQuery.js:
```diff
@@ -105,6 +113,16 @@ var MediaQuery = {
+   * @function
+   * @private
+   */
+  _reinit() {
```
`_reInit()` (with an uppercase)

SassNinja replied:
done

ncoden commented on js/foundation.util.mediaQuery.js:
```diff
@@ -105,6 +113,16 @@ var MediaQuery = {
   },
 
+  /**
+   * Reinitializes the media query helper what might be necessary if your CSS is not (completely) loaded while _init() is executed.
```
```
Reinitializes the media query helper.
Useful if your CSS breakpoint configuration has just been loaded or has changed since the initialization.
```

SassNinja committed:
fix: reset queries in MediaQuery _init method & adjust naming

This is necessary because using _reInit would populate the queries with duplicates again otherwise.

ncoden reviewed (approved):
LGTM! Thank you @SassNinja 😄

aj1991 commented:
Nice one 👍
